open-machine-learning-jupyter-book/assignments/llm/basic/transformer-architecture.ipynb

```diff
@@ -0,0 +1,726 @@
+    "        x2 = self.norm_3(x)\n",
+    "        x = x + self.dropout_3(self.ff(x2))\n",
+    "        return x\n",
+    "\n",
+    "class Decoder(nn.Module):\n",
+    "    def __init__(self, vocab_size, d_model, N, heads, dropout):\n",
+    "        super().__init__()\n",
+    "        self.N = N\n",
+    "        self.embed = Embedder(vocab_size, d_model)\n",
+    "        self.pe = PositionalEncoder(d_model, dropout=dropout)\n",
+    "        self.layers = get_clones(DecoderLayer(d_model, heads, dropout), N)\n",
+    "        self.norm = Norm(d_model)\n",
+    "\n",
+    "    def forward(self, trg, e_outputs, src_mask, trg_mask):\n",
+    "        x = self.embed(trg)\n",
+    "        x = self.pe(x)\n",
+    "        for i in range(self.N):\n",
+    "            x = self.layers[i](x, e_outputs, src_mask, trg_mask)\n",
+    "        return self.norm(x)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In these implementations:\n",
+    "\n",
+    "The EncoderLayer and DecoderLayer classes define encoder and decoder layers, respectively.\n",
+    "The Encoder and Decoder classes define encoder and decoder modules, respectively, composed of multiple layers of encoder or decoder layers.\n",
+    "These classes follow the architecture described in the text, including the use of multi-head attention, feedforward layers, residual connections, and layer normalization."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The overall implementation of the Transformer encoder and decoder structure:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import torch\n",
+    "import torch.nn as nn\n",
+    "import torch.nn.functional as F\n",
+    "import time\n",
+    "import numpy as np\n",
+    "\n",
+    "class Transformer(nn.Module):\n",
+    "    def __init__(self, src_vocab, trg_vocab, d_model, N, heads, dropout):\n",
+    "        super().__init__()\n",
+    "        self.encoder = Encoder(src_vocab, d_model, N, heads, dropout)\n",
+    "        self.decoder = Decoder(trg_vocab, d_model, N, heads, dropout)\n",
+    "        self.out = nn.Linear(d_model, trg_vocab)\n",
+    "\n",
+    "    def forward(self, src, trg, src_mask, trg_mask):\n",
+    "        e_outputs = self.encoder(src, src_mask)\n",
+    "        d_output = self.decoder(trg, e_outputs, src_mask, trg_mask)\n",
+    "        output = self.out(d_output)\n",
+    "        return output\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The training process for the Transformer model:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Model parameters\n",
+    "d_model = 512\n",
+    "heads = 8\n",
+    "N = 6\n",
+    "src_vocab = len(EN_TEXT.vocab)\n",
```

Comment:
NameError: name 'EN_TEXT' is not defined

Reply:
fix